vms/evm/warp/backend.go

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/libevm/log"
+	"github.com/ava-labs/libevm/metrics"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p/acp118"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/vms/evm/uptimetracker"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp/message"
+	"github.com/ava-labs/avalanchego/vms/platformvm/warp/payload"
+)
+
+const (
+	ParseErrCode = iota + 1
+	VerifyErrCode
+)
+
+var (
+	_ acp118.Verifier = (*Verifier)(nil)
+
+	ErrValidateBlock     = errors.New("failed to validate block message")
+	ErrVerifyWarpMessage = errors.New("failed to verify warp message")
+)
+
+// BlockStore provides access to accepted blocks.
+type BlockStore interface {
+	GetBlock(ctx context.Context, blockID ids.ID) (snowman.Block, error)
+}
+
+// DB stores and retrieves warp messages from the underlying database.
+type DB struct {
+	db database.Database
+}
+
+// NewDB creates a new warp message database.
+func NewDB(db database.Database) *DB {
+	return &DB{
+		db: db,
+	}
+}
+
+// Add stores a warp message in the database and cache.
+func (d *DB) Add(unsignedMsg *warp.UnsignedMessage) error {
+	msgID := unsignedMsg.ID()
+	log.Debug("Adding warp message to backend", "messageID", msgID)
```

Comment:
If we want this logging - I'm open to having it in the call-site at the rpc layer. I think we added these logs to increase observability, but I don't think that more logs == better. We already have [this](https://github.com/ava-labs/coreth/blob/0b315e0b3cb83d09a6422a7589ecfc920f68cbda/precompile/contracts/warp/config.go#L124) log for when we process a warp event so adding more logging into the database (which conceptually is already a weird place to add logging to) around this is just adding noise.